### PR TITLE
Generate valid Gaussian states, unitaries, and channels

### PR DESCRIPTION
**Pull request by apkille**
As discussed in #9, this PR generates covariance matrices in `randstate` and transformations in `randunitary` and `randchannel` with symplectic diagonalization discussed, e.g., in https://journals.aps.org/rmp/pdf/10.1103/RevModPhys.84.621.

### CODE REVIEW

File: src/randoms.jl [hunk 1]
```diff
@@ -1,43 +1,130 @@
...
+end
+function randsymplectic(::Type{T}, nmodes::N; passive = false) where {T, N<:Int} 
+    symp = randsymplectic(nmodes, passive)
+    return T(symp)
+end
+
+# Generates unitary matrix randomly distributed over Haar measure;
+# see https://arxiv.org/abs/math-ph/0609050 for algorithm.
+# This approach is much faster than using rand(Haar(2), nmodes) from RandomMatrices.jl
+function _rand_unitary(nmodes::N) where {N<:Int}
```
Comment (akirakyle) [inline src/randoms.jl:111]: Is this so much faster that it might be worth changing the implementation in QuantumOpticsBase.jl to match?

Comment (apkille) [reply src/randoms.jl:111]: Yes, I would say so:
```
julia> @benchmark rand(Haar(2), 20)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  22.625 μs …   8.383 ms  ┊ GC (min … max):  0.00% … 99.19%
 Time  (median):     27.250 μs               ┊ GC (median):     0.00%
 Time  (mean ± σ):   31.482 μs ± 151.002 μs  ┊ GC (mean ± σ):  12.35% ±  2.79%

                ▂▄▂▆█▆▄▁                                        
  ▁▃▂▂▂▂▂▂▂▂▂▃▄▆█████████▇▅▄▄▃▄▄▃▃▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▃
  22.6 μs         Histogram: frequency by time         37.2 μs <

 Memory estimate: 58.45 KiB, allocs estimate: 41.

julia> @benchmark Gabs._rand_unitary(10)
BenchmarkTools.Trial: 10000 samples with 5 evaluations.
 Range (min … max):  6.917 μs …  1.210 ms  ┊ GC (min … max): 0.00% … 98.92%
 Time  (median):     7.342 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   8.082 μs ± 20.767 μs  ┊ GC (mean ± σ):  6.53% ±  3.42%

   ▅▇   ▁▁▄█▆▇                                                
  ▃██▇█▆███████▇▄▂▃▂▂▂▃▃▃▄▄▆▆▆▆▅▆▅▆▄▅▃▃▄▃▃▂▂▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁ ▃
  6.92 μs        Histogram: frequency by time        9.06 μs <

 Memory estimate: 14.92 KiB, allocs estimate: 19.
```
Plus, RandomMatrices needs a lot of love on the maintenance side, so I'm hesitant to add it as a dependency for just one feature.

Comment (akirakyle) [reply src/randoms.jl:111]: Fair, something to keep in mind though if QuantumOpticsBase ever ends up as a dependency for Gabs

**Comment by apkille** [comment]
@akirakyle the implementation of `randchannel` in this PR breaks the Gaussian CP map condition `N + iΩ ≥ iTΩTᵀ`:
```
julia> nmodes = rand(1:20);

julia> rc = randchannel(nmodes);

julia> Omega = symplecticform(nmodes);

julia> @test all(i -> ((i >= 0) || isapprox(i, 0.0, atol = 1e-5)), real(eigvals(rc.noise .+ im*Omega .- im*rc.transform*Omega*rc.transform')))
Test Failed at REPL[15]:1
  Expression: all((i->begin
            #= REPL[15]:1 =#
            i >= 0 || isapprox(i, 0.0, atol = 1.0e-5)
        end), real(eigvals((rc.noise .+ im * Omega) .- im * rc.transform * Omega * (rc.transform)')))
```
 Any idea why this is? I checked my work with Serafini's discussion on pgs. 98 and 99 and it looks fine to me.

File: src/randoms.jl (hunk 1 above)
Comment (apkille) [reply src/randoms.jl:111]: I agree. The QO maintainers might be interested in removing RandomMatrices as a dependency, I know they are working in reducing the slow startup time (you may have already seen this issue but I am putting it here anyways: https://github.com/qojulia/QuantumOptics.jl/issues/417).

**Comment by akirakyle** [comment]
Hmmm, staring at the code of `randchannel` I don't see anything obvious. The checks for randsymplectic are passing?

I think one observation which would simplify the code is that it is actually not necessary to have an environment covariance matrix other than identity if the goal is to generate random channels since a random unitary on 2n modes corresponds to a random channel on n modes after tracing out the other n modes assuming they are in vacuum. Physically the intuition behind this is that noise on the n mode subsystem will arise due to two-mode squeezing with the environment as tracing out one mode of such a state reveals a thermal state.

Perhaps try simplifying the code with this observation first then see if the checks are failing.

**Comment by apkille** [comment]
The issue was that I derived transform and noise matrices for `randchannel` in the canonical basis. However, my tests were with symplectic forms in the block diagonal basis, which were generated with `symplecticform`. That's my bad. Given that most papers on Gaussian quantum info use the canonical basis, I will make a breaking change in this PR that changes the matrix generated by `symplecticform` to be that in the canonical basis, e.g.,
```
julia> symplecticform(2)
4×4 Matrix{Float64}:
  0.0  1.0   0.0  0.0
 -1.0  0.0   0.0  0.0
  0.0  0.0   0.0  1.0
  0.0  0.0  -1.0  0.0
```
All the tests pass now. I should make a point about symplectic basis in the documentation soon.

> Perhaps try simplifying the code with this observation first then see if the checks are failing.

This note was still very helpful and was used to simplify the internals behind `randchannel`. Thanks!

**Comment by akirakyle** [comment]
Ah yes the explicit representation of the symplectic form is perhaps the most important thing as it dictates the representation of all other objects in the Gaussian formalism! While the representation you're calling "canonical" is the one used by Weedbrook and Serafini and seemingly most common among QI theorists these days, one often encounters the "sigma z" representation given by Serafini Eq. 3.7 in more engineering focused contexts as that representation is closely related to the classical scattering matrix response a linear time-translation-invariant system. For example see this paper https://arxiv.org/abs/2410.09976. I'm actually quite interested in also having ways model the kind of frequency selective gaussian operations that that paper discusses and such systems can be also be described by the symplectic group, just either with different representation or reinterpretation of the physical nature of the operators.

I think while developing the functionality of this library it's good to fix the symplectic representation to avoid confusion such as the one you encountered, but long term it would be really helpful to support all the different representations with functions to convert between them!

**Comment by akirakyle** [comment]
Yeah something like that seems like a good, julian approach to me!

**Pull request merged**